src/main/scala/xiangshan/mem/vector/VSFlowQueue.scala

```diff
@@ -199,6 +201,86 @@ class VsFlowQueueIOBundle(implicit p: Parameters) extends VLSUBundle {
   // MMIO flows execution path
   val uncacheOutstanding = Input(Bool())
   val uncache = new UncacheWordIO
+
+  // update tval when exception happens
+  // val exceptionAddrValid = Output(Bool())
+  val exceptionAddr = new ExceptionAddrIO
+}
+
+class VsExceptionBuffer(implicit p: Parameters) extends VLSUModule with HasCircularQueuePtrHelper {
+  val io = IO(new Bundle() {
+    val redirect = Flipped(ValidIO(new Redirect))
+    val flowWriteback = Vec(VecStorePipelineWidth, Flipped(ValidIO(new VecStoreExuOutput())))
+    val exceptionAddr = new ExceptionAddrIO
+  })
+
+  val req_valid = RegInit(false.B)
+  val req = Reg(new VecStoreExuOutput())
+
+  // enqueue
+  // S1:
+  val s1_req = VecInit(io.flowWriteback.map(_.bits))
+  val s1_valid = VecInit(io.flowWriteback.map(_.valid))
+  
+  // S2: delay 1 cycle
+  val s2_req = RegNext(s1_req)
+  val s2_valid = (0 until VecStorePipelineWidth).map(i =>
+    RegNext(s1_valid(i)) &&
+    !s2_req(i).uop.robIdx.needFlush(RegNext(io.redirect)) && 
+    !s2_req(i).uop.robIdx.needFlush(io.redirect)
+  )
+  val s2_has_exception = s2_req.map(x => ExceptionNO.selectByFu(x.uop.exceptionVec, VstuCfg).asUInt.orR)
+
+  val s2_enqueue = Wire(Vec(VecStorePipelineWidth, Bool()))
+  for (w <- 0 until VecStorePipelineWidth) {
+    s2_enqueue(w) := s2_valid(w) && s2_has_exception(w)
+  }
+
+  when (req_valid && req.uop.robIdx.needFlush(io.redirect)) {
+    req_valid := s2_enqueue.asUInt.orR
+  }.elsewhen (s2_enqueue.asUInt.orR) {
+    req_valid := req_valid || true.B
+  }
+
+  def selectOldest[T <: VecStoreExuOutput](valid: Seq[Bool], bits: Seq[T]): (Seq[Bool], Seq[T]) = {
+    assert(valid.length == bits.length)
+    if (valid.length == 0 || valid.length == 1) {
+      (valid, bits)
+    } else if (valid.length == 2) {
+      val res = Seq.fill(2)(Wire(Valid(chiselTypeOf(bits(0)))))
+      for (i <- res.indices) {
+        res(i).valid := valid(i)
+        res(i).bits := bits(i)
+      }
+      val oldest = Mux(
+        valid(0) && valid(1),
+        Mux(
+          Cat(bits(0).segmentIdx, bits(0).fieldIdx) < Cat(bits(1).segmentIdx, bits(1).fieldIdx),
+          res(0), res(1)
+        ),
+        Mux(valid(0), res(0), res(1))
+      )
+      (Seq(oldest.valid), Seq(oldest.bits))
+    } else {
+      val left = selectOldest(valid.take(valid.length / 2), bits.take(bits.length / 2))
+      val right = selectOldest(valid.takeRight(valid.length - (valid.length / 2)), bits.takeRight(bits.length - (bits.length / 2)))
+      selectOldest(left._1 ++ right._1, left._2 ++ right._2)
+    }
+  }
```

Comment:
Just complain.
Is there any more general selectOldest function? We always create a new selectOldest function everytime age comparasion is needed.

Reply:
Need no modification